ttkwidgets/notebook.py

```diff
@@ -0,0 +1,926 @@
+        :param compound: how the tab label and image are organized
+        :param sticky: for the widget inside the notebook
+        :param padding: padding (int) around the widget in the notebook
+        :param state: state ('normal' or 'disabled') of the tab
+        """
+        # Todo: underline
+        name = str(widget)
+        if name in self._indexes:
+            ind = self._indexes[name]
+            self.tab(ind, **kwargs)
+            self._show(ind)
+            self.update_idletasks()
+        else:
+            sticky = kwargs.pop('sticky', 'ewns')
+            padding = kwargs.pop('padding', 0)
+            self._tabs[self._nb_tab] = widget
+            ind = self._nb_tab
+            self._indexes[name] = ind
+            self._tab_labels[ind] = Tab(self._tab_frame2, tab_nb=ind,
+                                        closecommand=self._closecommand,
+                                        closebutton=self._closebutton,
+                                        **kwargs)
+            self._tab_labels[ind].bind('<ButtonRelease-1>', self._on_click)
+            self._tab_labels[ind].bind('<ButtonRelease-3>', lambda e: self._popup_menu(e, ind))
+            self._tab_labels[ind].bind('<ButtonPress-1>', lambda e: self._on_press(e, ind))
+            self._body.configure(height=max(self._body.winfo_height(), widget.winfo_reqheight()),
+                                 width=max(self._body.winfo_width(), widget.winfo_reqwidth()))
+
+            self._tab_options[ind] = dict(text='', image='', compound='none', state='normal')
+            self._tab_options[ind].update(kwargs)
+            self._tab_options[ind].update(dict(padding=padding, sticky=sticky))
+            self._tab_menu_entries[ind] = self._tab_menu.index('end')
+            self._tab_list.state(['!disabled'])
+            self._active_tabs.append(ind)
+            self._show(self._nb_tab, new=True, update=True)
+
+            self._nb_tab += 1
+            self._menu_insert(ind, kwargs.get('text', ''))
+        return ind
+
+    def insert(self, where, widget, **kwargs):
+        """
+        Insert WIDGET at the position given by WHERE in the notebook.
+
+        For keyword options, see add method.
+        """
+        existing = str(widget) in self._indexes
+        index = self.add(widget, **kwargs)
+        if where == 'end':
+            if not existing:
+                return
+        where = self.index(where)
+        self._visible_tabs.remove(index)
+        self._visible_tabs.insert(where, index)
+        for i in range(where, len(self._visible_tabs)):
+            ind = self._visible_tabs[i]
+            self._tab_labels[ind].grid_configure(column=i)
+        self.update_idletasks()
+        self._on_configure()
+
+    def enable_traversal(self):
+        self.bind('<Control-Tab>', lambda e: self.select_next(True))
+        self.bind('<Shift-Control-ISO_Left_Tab>', lambda e: self.select_prev(True))
+
+    def index(self, tab_id):
```

Comment:
While working on `test_notebook_insert`, I noticed a weird thing about this function. It does not return the index as it is visible to the user, rather is simply returns a valid `tab_id` given a host of different inputs. In fact, it does not even do this consistently: `tk.END` yields the length of `self._tabs`, which is not even a valid tab id or tab index!

I think the behaviour of this function must be consistent with that of `ttk.Notebook`, that is, it must return the index of the tab as it is visible to the user. To do this, it should use `self._visible_tabs`. As this function is used in a lot of other functions in the whole class, changing its behaviour completely would be ill-advised. Therefore, I think changing this function to be `_index()` and creating a new function `index()` is best. I have done this [here](https://github.com/TkinterEP/ttkwidgets/tree/notebook_index).